client/dns_client.mli

```diff
@@ -78,8 +79,9 @@ sig
 
   type t
 
-  val create : ?size:int -> ?rng:(int -> Cstruct.t) -> ?nameserver:T.ns_addr -> clock:(unit -> int64) -> T.stack -> t
-  (** [create ~size ~rng ~nameserver ~clock stack] creates the state of the DNS client. *)
+  val create : ?size:int -> ?rng:(int -> Cstruct.t) -> ?nameserver:T.ns_addr -> clock:(unit -> int64) -> ?timeout:int64 -> T.stack -> t
+  (** [create ~size ~rng ~nameserver ~clock ~timeout stack] creates the state of the DNS client.
+      We use this [timeout] as both connect and request timeout. *)
```

Comment:
I agree that the documentation string can be extended. An attempt to answer your questions above:
- there is no timeout set for send, since the semantics of send AFAICT are to immediately return (i.e. does not wait for the packet being sent / the TCP ACK being received).
- on Unix sockets API it is SO_RCVTIMEO, and the time we wait in receive. from an DNS client user perspective, it is the "timeout until the request is discarded" (I don't really have a strong opinion whether this should be `request` or `response`, @linse said it is usually named `request timeout` if I understand correctly).
- document measurement unit: yes, indeed this needs to be more precise
- `polymorphic error` I agree that this is useful, in the future, especially when thinking about UDP where we'd likely want to retransmit the question multiple times. But I suspect it is a slightly more involved change that we should address separately.
- maybe we should revise and have separate `connect` and `request` (or `response` if we settle on that) timeout. we also discussed whether they should be passable (as optional labelled argument) for each client request (in addition to the global value for `create`) -- similar to `?nameserver` -- what do you think?

Reply:
- The semantics of `write`/`sendmsg` depends on whether or not the socket is in non-blocking mode (that's the case for `dns-client.lwt` and *not* for `dns-client.unix`). See https://linux.die.net/man/2/sendmsg:
    > When the message does not fit into the send buffer of the socket, send() normally blocks, unless the socket has been placed in nonblocking I/O mode. In nonblocking mode it would fail with the error EAGAIN or EWOULDBLOCK in this case. The select(2) call may be used to determine when it is possible to send more data.

   (So `write`/`sendto`/`sendmsg` doesn't wait for packet sent / TCP ack, but it does block if the output buffer is full. I believe that for TCP the send window also plays into this, but I have not checked.)
- I agree that this is normally called `request timeout`, but in those cases it's a timeout for the whole operation ("timeout for how long it can take to send a request and (start) receiving a response", or "timeout for how long it can take to send and receive a response"). As noted I think such a timeout for the entire operation is more useful to end-users than low-level timeouts on each I/O operation. ie "if I haven't gotten a DNS reply within 20 seconds, try again" - the user doesn't care how many times `recv` was called within the DNS stack. In any case my point here is that we can do it however we like, but I think we should document clearly what the user should expect.
- Re: separate connect / request timeouts: I don't have an opinion, I guess you might want smaller timeout on `connect` to a local resolver than for a recursive request/response cycle. Where it might matter is when we need to do a TCP reconnect in order to get a response/response shot off - in that case I guess the timeout should cover all of that instead of each individual step (or we should document that when using TCP it might take twice as long for the timeout to kick in).

Reply:
We addressed this by introducing a budget, implementing a timeout for send, and added an example on how to specify a timeout.